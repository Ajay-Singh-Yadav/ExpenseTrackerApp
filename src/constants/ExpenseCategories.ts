  export const expenseCategories = [
    { label: 'Food & Drinks', value: 'Food & Drinks' },
    { label: 'Travel', value: 'Travel' },
    { label: 'Loan EMIs', value: 'Loan EMIs' },
    { label: 'Rent', value: 'Rent' },
    { label: 'Entertainment', value: 'Entertainment' },
    { label: 'Bills', value: 'Bills' },
    { label: 'Health Care', value: 'Health Care' },
    { label: 'Shopping', value: 'Shopping' },
    { label: 'Vacation', value: 'Vacation' },
    { label: 'Subscriptions', value: 'Subscriptions' },
    { label: 'Others', value: 'Others' },
  ];