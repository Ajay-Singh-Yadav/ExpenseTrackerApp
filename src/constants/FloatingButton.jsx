import React from 'react';
import { TouchableOpacity, StyleSheet } from 'react-native';
import Entypo from 'react-native-vector-icons/Entypo';
import { scale } from 'react-native-size-matters';

const FloatingButton = React.memo(({ onPress }) => (
  <TouchableOpacity style={styles.addButton} onPress={onPress}>
    <Entypo name="plus" size={35} color="white" />
  </TouchableOpacity>
));

export default FloatingButton;

const styles = StyleSheet.create({
  addButton: {
    backgroundColor: '#4CAF50',
    borderRadius: scale(60),
    width: scale(50),
    height: scale(50),
    justifyContent: 'center',
    alignItems: 'center',
    position: 'absolute',
    right: scale(30),
    bottom: scale(30),
    zIndex: 10,
  },
});
