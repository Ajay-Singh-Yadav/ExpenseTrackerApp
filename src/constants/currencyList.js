export const currencyList = [
  {
    code: 'USD',
    name: 'United States Dollar',
    icon: 'dollar-sign',
    iconLibrary: 'FontAwesome5',
  },
  {
    code: 'INR',
    name: 'Indian Rupee',
    icon: 'currency-inr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  { code: 'EUR', name: 'Euro', icon: 'euro-sign', iconLibrary: 'FontAwesome5' },
  {
    code: 'GBP',
    name: 'British Pound',
    icon: 'pound-sign',
    iconLibrary: 'FontAwesome5',
  },
  {
    code: 'JPY',
    name: 'Japanese Yen',
    icon: 'yen-sign',
    iconLibrary: 'FontAwesome5',
  },
  {
    code: 'AUD',
    name: 'Australian Dollar',
    icon: 'currency-usd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'CAD',
    name: 'Canadian Dollar',
    icon: 'currency-usd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'CHF',
    name: 'Swiss Franc',
    icon: 'currency-chf',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'CNY',
    name: 'Chinese Yuan',
    icon: 'currency-cny',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'SEK',
    name: 'Swedish Krona',
    icon: 'currency-sek',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'NZD',
    name: 'New Zealand Dollar',
    icon: 'currency-nzd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'SGD',
    name: 'Singapore Dollar',
    icon: 'currency-sgd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'HKD',
    name: 'Hong Kong Dollar',
    icon: 'currency-hkd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'NOK',
    name: 'Norwegian Krone',
    icon: 'currency-nok',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'KRW',
    name: 'South Korean Won',
    icon: 'currency-krw',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'TRY',
    name: 'Turkish Lira',
    icon: 'currency-try',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'RUB',
    name: 'Russian Ruble',
    icon: 'currency-rub',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'ZAR',
    name: 'South African Rand',
    icon: 'currency-zar',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'BRL',
    name: 'Brazilian Real',
    icon: 'currency-brl',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'MXN',
    name: 'Mexican Peso',
    icon: 'currency-mxn',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'DKK',
    name: 'Danish Krone',
    icon: 'currency-dkk',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'PLN',
    name: 'Polish Zloty',
    icon: 'currency-pln',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'THB',
    name: 'Thai Baht',
    icon: 'currency-thb',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'MYR',
    name: 'Malaysian Ringgit',
    icon: 'currency-myr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'IDR',
    name: 'Indonesian Rupiah',
    icon: 'currency-idr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'PHP',
    name: 'Philippine Peso',
    icon: 'currency-php',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'CZK',
    name: 'Czech Koruna',
    icon: 'currency-czk',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'HUF',
    name: 'Hungarian Forint',
    icon: 'currency-huf',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'ILS',
    name: 'Israeli Shekel',
    icon: 'currency-ils',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'AED',
    name: 'UAE Dirham',
    icon: 'currency-aed',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'SAR',
    name: 'Saudi Riyal',
    icon: 'currency-sar',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'QAR',
    name: 'Qatari Riyal',
    icon: 'currency-qar',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'EGP',
    name: 'Egyptian Pound',
    icon: 'currency-egp',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'NGN',
    name: 'Nigerian Naira',
    icon: 'currency-ngn',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'BDT',
    name: 'Bangladeshi Taka',
    icon: 'currency-bdt',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'PKR',
    name: 'Pakistani Rupee',
    icon: 'currency-pkr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'LKR',
    name: 'Sri Lankan Rupee',
    icon: 'currency-lkr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'KWD',
    name: 'Kuwaiti Dinar',
    icon: 'currency-kwd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'BHD',
    name: 'Bahraini Dinar',
    icon: 'currency-bhd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'OMR',
    name: 'Omani Rial',
    icon: 'currency-omr',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'MAD',
    name: 'Moroccan Dirham',
    icon: 'currency-mad',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'JOD',
    name: 'Jordanian Dinar',
    icon: 'currency-jod',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'VND',
    name: 'Vietnamese Dong',
    icon: 'currency-vnd',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'ARS',
    name: 'Argentine Peso',
    icon: 'currency-ars',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'CLP',
    name: 'Chilean Peso',
    icon: 'currency-clp',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'COP',
    name: 'Colombian Peso',
    icon: 'currency-cop',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'UAH',
    name: 'Ukrainian Hryvnia',
    icon: 'currency-uah',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'ISK',
    name: 'Icelandic Krona',
    icon: 'currency-isk',
    iconLibrary: 'MaterialCommunityIcons',
  },
  {
    code: 'KES',
    name: 'Kenyan Shilling',
    icon: 'currency-kes',
    iconLibrary: 'MaterialCommunityIcons',
  },
];
