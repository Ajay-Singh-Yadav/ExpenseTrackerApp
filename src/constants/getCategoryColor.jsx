const getCategoryColor = (type, category) => {
  if (type.toLowerCase() === 'income') return '#22c55e';

  switch (category) {
    case 'Food & Drinks':
      return '#f43f5e';
    case 'Travel':
      return '#fc034e';
    case 'Entertainment':
      return '#fc5203';
    case 'Loan EMIs':
      return '#fdba74';
    case 'Rent':
      return '#6203fc';
    case 'Bills':
      return '#03fc52';
    case 'Health Care':
      return '#03fcfc';
    case 'Shopping':
      return '#fc03fc';
    case 'Vacation':
      return '#87CEEB';
    case 'Subscriptions':
      return '#FF6F00';
    default:
      return '#95A5A6';
  }
};

export default getCategoryColor;
