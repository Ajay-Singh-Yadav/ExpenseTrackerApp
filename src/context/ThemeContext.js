import React, {
  createContext,
  useContext,
  useEffect,
  useMemo,
  useState,
} from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

// Create the ThemeContext
const ThemeContext = createContext();

const loadThemeByKey = async key => {
  switch (key) {
    case 'coffee':
      return (await import('../themes/coffee')).default;
    case 'dark':
      return (await import('../themes/dark')).default;
    case 'forest':
      return (await import('../themes/forest')).default;
    case 'purple':
      return (await import('../themes/purple')).default;
    case 'ocean':
      return (await import('../themes/ocean')).default;
    default:
      return (await import('../themes/coffee')).default;
  }
};

export const ThemeProvider = ({ children }) => {
  const [theme, setTheme] = useState(require('../themes/coffee').default);
  const [currentTheme, setCurrentTheme] = useState('coffee');
  const [isLoading, setIsLoading] = useState(true);

  const changeTheme = async key => {
    setIsLoading(true);
    const newTheme = await loadThemeByKey(key);
    setTheme(newTheme);
    setCurrentTheme(key);
    await AsyncStorage.setItem('appTheme', key);
    setIsLoading(false);
  };

  useEffect(() => {
    (async () => {
      const savedTheme = await AsyncStorage.getItem('appTheme');
      const themeKey = savedTheme || 'coffee';
      const loadedTheme = await loadThemeByKey(themeKey);
      setTheme(loadedTheme);
      setCurrentTheme(themeKey);
      setIsLoading(false);
    })();
  }, []);

  const value = useMemo(
    () => ({
      theme,
      currentTheme,
      setTheme: changeTheme,
      isLoading,
    }),
    [theme, currentTheme, isLoading],
  );

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
};

export const useTheme = () => useContext(ThemeContext);
