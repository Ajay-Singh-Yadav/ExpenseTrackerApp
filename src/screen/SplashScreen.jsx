import React, { memo, useEffect, useRef } from 'react';
import { Animated, View, Image, InteractionManager, Text } from 'react-native';
import useSplash from '../hooks/useSplash';
import { useNavigation } from '@react-navigation/native';
import LottieView from 'lottie-react-native';
// import getAuth from '@react-native-firebase/auth';
import { getAuth, onAuthStateChanged } from '@react-native-firebase/auth';

const SplashScreen = () => {
  const navigation = useNavigation();
  const styles = useSplash();
  const fadeAnim = useRef(new Animated.Value(0)).current;
  const lottieRef = useRef(null);

  useEffect(() => {
    const auth = getAuth();
    Animated.timing(fadeAnim, {
      toValue: 1,
      duration: 900,
      useNativeDriver: true,
    }).start();

    const chechUser = () => {
      const unsubscribe = onAuthStateChanged(auth, user => {
        if (user) {
          navigation.replace('Home');
        } else {
          navigation.replace('LogIn');
        }
      });
      return unsubscribe;
    };

    const timeout = setTimeout(chechUser, 2500);
    return () => {
      clearTimeout(timeout);
    };
  }, [navigation]);

  useEffect(() => {
    lottieRef.current?.play();

    return () => {
      lottieRef.current?.reset();
    };
  }, []);

  return (
    <Animated.View style={[styles.container, { opacity: fadeAnim }]}>
      <LottieView
        ref={lottieRef}
        source={require('../assets/animations/Fintech.json')}
        autoPlay
        loop={true}
        style={styles.lottie}
      />
    </Animated.View>
  );
};

export default memo(SplashScreen);
