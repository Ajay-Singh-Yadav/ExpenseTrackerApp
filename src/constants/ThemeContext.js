import React, {
  createContext,
  useState,
  useEffect,
  useContext,
  useCallback,
  useMemo,
} from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import { Themes } from '../constants/theme';

const ThemeContext = createContext();

export const ThemeProvider = ({ children }) => {
  const [themeKey, setThemeKey] = useState('dark');
  const [theme, setTheme] = useState(Themes.dark);

  // Load saved theme on mount
  useEffect(() => {
    (async () => {
      try {
        const storedTheme = await AsyncStorage.getItem('appTheme');
        if (storedTheme && Themes[storedTheme]) {
          setThemeKey(storedTheme);
          setTheme(Themes[storedTheme]);
        }
      } catch (error) {
        console.error('Error loading theme:', error);
      }
    })();
  }, []);

  const toggleTheme = useCallback(async key => {
    if (Themes[key]) {
      setThemeKey(key);
      setTheme(Themes[key]);
      await AsyncStorage.setItem('appTheme', key);
    }
  }, []);

  const value = useMemo(
    () => ({
      theme,
      themeKey,
      toggleTheme,
    }),
    [theme, themeKey, toggleTheme],
  );

  return (
    <ThemeContext.Provider value={value}>{children}</ThemeContext.Provider>
  );
};

export const useTheme = () => useContext(ThemeContext);
