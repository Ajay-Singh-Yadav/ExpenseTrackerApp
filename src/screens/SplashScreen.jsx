import React, { useMemo } from 'react';
import { View, StyleSheet } from 'react-native';

import FastImage from 'react-native-fast-image';
import { moderateScale } from 'react-native-size-matters';

import { useTheme } from '../constants/ThemeContext';

const welcomeImage = require('../assets/images/welcome.png');

const SplashScreen = () => {
  const { theme } = useTheme();

  const styles = useMemo(
    () =>
      StyleSheet.create({
        container: {
          flex: 1,
          justifyContent: 'center',
          alignItems: 'center',
          backgroundColor: theme.bgColor,
        },
        logo: {
          width: moderateScale(150),
          height: moderateScale(150),
          resizeMode: 'contain',
        },
      }),
    [theme],
  );

  return (
    <View style={styles.container}>
      <FastImage
        source={welcomeImage}
        style={styles.logo}
        testID="splash-logo"
      />
    </View>
  );
};

export default React.memo(SplashScreen);
