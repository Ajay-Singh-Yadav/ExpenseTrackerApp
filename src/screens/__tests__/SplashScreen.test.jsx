import React from 'react';
import { render } from '@testing-library/react-native';
import SplashScreen from '../SplashScreen';
import { useTheme } from '../../constants/ThemeContext';
import FastImage from 'react-native-fast-image';

jest.mock('../../constants/ThemeContext', () => ({
  useTheme: jest.fn(),
}));

describe('SplashScreen', () => {
  const mockTheme = { bgColor: '#ffffff' };

  beforeEach(() => {
    useTheme.mockReturnValue({ theme: mockTheme });
  });

  it('renders without crashing', () => {
    render(<SplashScreen />);
  });

  it('matches snapshot', () => {
    const tree = render(<SplashScreen />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders FastImage with correct testID', () => {
    const { getByTestId } = render(<SplashScreen />);
    expect(getByTestId('splash-logo')).toBeTruthy();
  });

  it('applies correct background color from theme', () => {
    const { getByTestId, toJSON } = render(<SplashScreen />);
    const json = toJSON();
    expect(json.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ backgroundColor: '#ffffff' }),
      ]),
    );
  });

  it('renders logo with correct size', () => {
    const { getByTestId } = render(<SplashScreen />);
    const logo = getByTestId('splash-logo');
    expect(logo.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ width: 140, height: 140 }),
      ]),
    );
  });

  it('re-renders when theme changes', () => {
    const { rerender, toJSON } = render(<SplashScreen />);
    useTheme.mockReturnValue({ theme: { bgColor: '#000000' } });
    rerender(<SplashScreen />);
    const json = toJSON();
    expect(json.props.style).toEqual(
      expect.arrayContaining([
        expect.objectContaining({ backgroundColor: '#000000' }),
      ]),
    );
  });

  it('uses React.memo to prevent unnecessary re-render', () => {
    const { rerender } = render(<SplashScreen />);
    const firstRender = render(<SplashScreen />);
    rerender(<SplashScreen />);
    // If memo works, component tree should be structurally same
    expect(firstRender.toJSON()).toEqual(firstRender.toJSON());
  });
});
