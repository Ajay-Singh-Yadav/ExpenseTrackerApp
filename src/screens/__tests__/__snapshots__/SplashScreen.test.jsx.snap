// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SplashScreen matches snapshot 1`] = `
<View
  style={
    [
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "center",
      },
      {
        "backgroundColor": "#ffffff",
      },
    ]
  }
>
  <View
    style={
      [
        {
          "overflow": "hidden",
        },
        [
          {
            "height": 140,
            "resizeMode": "contain",
            "width": 140,
          },
        ],
      ]
    }
  >
    <FastImageView
      defaultSource={null}
      resizeMode="cover"
      source={
        {
          "testUri": "../../../src/assets/images/welcome.png",
        }
      }
      style={
        {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        }
      }
      testID="splash-logo"
    />
  </View>
</View>
`;
